R/deploymentTarget.R

```diff
@@ -1,93 +1,241 @@
-  accountDetails <- findAccount(account, server)
+  # No local deployment record. Get it from the server.
   application <- getApplication(accountDetails$name, accountDetails$server, appId)
 
-  createDeploymentTarget(
+  # Note: The account+server of this deployment record may
+  # not correspond to the original content creator.
+  deployment <- createDeploymentTarget(
     appName = application$name,
     appTitle = application$title %||% appTitle,
     appId = application$id,
-    envVars = NULL,
+    envVars = envVars,
     username = application$owner_username %||% accountDetails$name,
     account = accountDetails$name,
     server = accountDetails$server
   )
+
+  return(list(
+    accountDetails = accountDetails,
+    deployment = deployment
+  ))
+}
+
+# Discover the deployment target given appName.
+#
+# When appName is provided it identifies content previously created by a locally configured account.
+#
+# The account details from the deployment record identify the final credentials we will use, as
+# account+server may not have been specified by the caller.
+deploymentTargetFromAppName <- function(
+  recordPath = ".",
+  appName = NULL,
+  appTitle = NULL,
+  envVars = NULL,
+  account = NULL,
+  server = NULL,
+  forceUpdate = FALSE,
+  error_call = caller_env()
+) {
+
+  appDeployments <- deployments(
+    appPath = recordPath,
+    nameFilter = appName,
+    accountFilter = account,
+    serverFilter = server
+  )
+
+  # When the appName along with the (optional) account+server identifies exactly one previous
+  # deployment, use it.
+  if (nrow(appDeployments) == 1) {
+    deployment <- appDeployments[1, ]
+    deployment <- updateDeploymentTarget(deployment, appTitle, envVars)
+    accountDetails <- accountInfo(deployment$account, deployment$server)
+    return(list(
+      accountDetails = accountDetails,
+      deployment = deployment
+    ))
+  }
+
+  # When the appName identifies multiple targets, we may not have had an account+server constraint.
+  # Ask the user to choose.
+  if (nrow(appDeployments) > 1) {
+    deployment <- disambiguateDeployments(appDeployments, error_call = error_call)
+    deployment <- updateDeploymentTarget(deployment, appTitle, envVars)
+    accountDetails <- accountInfo(deployment$account, deployment$server)
+    return(list(
+      accountDetails = accountDetails,
+      deployment = deployment
+    ))
+  }
+
+  # When the appName does not identify a target, see if it exists on the server. That content is
+  # conditionally used. A resolved account is required.
+  accountDetails <- accountInfo(account, server)
+  if (!isPositCloudServer(accountDetails$server)) {
+    client <- clientForAccount(accountDetails)
+    application <- getAppByName(client, accountDetails, appName)
+    if (!is.null(application)) {
+      uniqueName <- findUnique(appName, application$name)
+      if (shouldUpdateApp(application, uniqueName, forceUpdate)) {
+        deployment <- createDeploymentTarget(
```

Comment:
I wonder if it would be useful to have an explicit function called that takes an application and an accountDetails and creates a deploymentTarget